plugins/SlicerT/SlicerT.cpp

```diff
@@ -0,0 +1,402 @@
+ * Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
+ * Boston, MA 02110-1301 USA.
+ *
+ */
+
+#include "SlicerT.h"
+
+#include <QDomElement>
+#include <cmath>
+#include <fftw3.h>
+
+#include "Engine.h"
+#include "InstrumentTrack.h"
+#include "PathUtil.h"
+#include "Song.h"
+#include "embed.h"
+#include "lmms_constants.h"
+#include "plugin_export.h"
+
+namespace lmms {
+
+extern "C" {
+Plugin::Descriptor PLUGIN_EXPORT slicert_plugin_descriptor = {
+	LMMS_STRINGIFY(PLUGIN_NAME),
+	"SlicerT",
+	QT_TRANSLATE_NOOP("PluginBrowser", "Basic Slicer"),
+	"Daniel Kauss Serna <daniel.kauss.serna@gmail.com>",
+	0x0100,
+	Plugin::Type::Instrument,
+	new PluginPixmapLoader("icon"),
+	nullptr,
+	nullptr,
+};
+} // end extern
+
+// ################################# SlicerT ####################################
+
+SlicerT::SlicerT(InstrumentTrack* instrumentTrack)
+	: Instrument(instrumentTrack, &slicert_plugin_descriptor)
+	, m_noteThreshold(0.6f, 0.0f, 2.0f, 0.01f, this, tr("Note threshold"))
+	, m_fadeOutFrames(400.0f, 0.0f, 8192.0f, 1.0f, this, tr("FadeOut"))
+	, m_originalBPM(1, 1, 999, this, tr("Original bpm"))
+	, m_sliceSnap(this, tr("Slice snap"))
+	, m_enableSync(false, this, tr("BPM sync"))
+	, m_originalSample()
+	, m_parentTrack(instrumentTrack)
+{
+	m_sliceSnap.addItem("Off");
+	m_sliceSnap.addItem("1/1");
+	m_sliceSnap.addItem("1/2");
+	m_sliceSnap.addItem("1/4");
+	m_sliceSnap.addItem("1/8");
+	m_sliceSnap.addItem("1/16");
+	m_sliceSnap.addItem("1/32");
+	m_sliceSnap.setValue(0);
+}
+
+void SlicerT::playNote(NotePlayHandle* handle, sampleFrame* workingBuffer)
+{
+	if (m_originalSample.frames() <= 1) { return; }
+
+	const int noteIndex = handle->key() - m_parentTrack->baseNote();
+	const fpp_t frames = handle->framesLeftForCurrentPeriod();
+	const f_cnt_t offset = handle->noteOffset();
+	const int bpm = Engine::getSong()->getTempo();
+	const float pitchRatio = 1 / pow(2, m_parentTrack->pitchModel()->value() / 1200);
+
+	float speedRatio = static_cast<float>(m_originalBPM.value()) / bpm;
+	if (!m_enableSync.value()) { speedRatio = 1; }
+	speedRatio *= pitchRatio;
+	speedRatio *= Engine::audioEngine()->processingSampleRate() / static_cast<float>(m_originalSample.sampleRate());
+
+	float sliceStart, sliceEnd;
+	if (noteIndex > m_slicePoints.size() - 2 || noteIndex < 0) // full sample if ouside range
+	{
+		sliceStart = 0;
+		sliceEnd = 1;
+	}
+	else
+	{
+		sliceStart = m_slicePoints[noteIndex];
+		sliceEnd = m_slicePoints[noteIndex + 1];
+	}
+
+	if (!handle->m_pluginData) { handle->m_pluginData = new PlaybackState(sliceStart); }
+
+	float noteDone = ((PlaybackState*)handle->m_pluginData)->noteDone();
+	float noteLeft = sliceEnd - noteDone;
+
+	if (noteLeft > 0)
+	{
+		int noteFrame = noteDone * m_originalSample.frames();
+
+		SRC_STATE* resampleState = ((PlaybackState*)handle->m_pluginData)->resamplingState();
+		SRC_DATA resampleData;
+		resampleData.data_in = (float*)(m_originalSample.data() + noteFrame);
```

Comment:
You should be using `static_cast<float*>` instead. This also applies for the line right after this one.

Reply:
I actually already tried that, and it gives this error:
`Static_cast from 'const sampleFrame *' (aka 'const array<float, DEFAULT_CHANNELS> *') to 'float *' is not allowed [bad_cxx_cast_generic]`
I also tried a few other casts and none seemed to work.

Reply:
Try making the cast `const`?

Reply:
Nope, and here are a few other casts that I tested:
const_cast:
`Const_cast from 'const sampleFrame *' (aka 'const array<float, DEFAULT_CHANNELS> *') to 'float *' is not allowed [bad_cxx_cast_generic]`
reinterpret_cast:
`Reinterpret_cast from 'const sampleFrame *' (aka 'const array<float, DEFAULT_CHANNELS> *') to 'float *' casts away qualifiers [bad_cxx_cast_qualifiers_away]`
dynamic_cast:
`'float' is not a class type [bad_dynamic_cast_not_class]`

Reply:
The array returns a `sampleFrame*`, which is an array with two float values, so after you call `data()` access the first element and get its address.

Reply:
Ok, this seems to work fine:
`&(m_originalSample.data() + noteFrame)[0][0]`

Reply:
Did you mean to deference the frame instead of using `[0]`? in any case, I think calling `data()` on the pointer to the frame is simpler.

Reply:
> Did you mean to deference the frame instead of using [0]? in any case, I think calling data() on the pointer to the frame is simpler.

Yes this is much simpler, done.